components/raftstore/src/store/peer_storage.rs

```diff
@@ -396,6 +396,38 @@ pub fn init_apply_state(engines: &Engines, region: &Region) -> Result<RaftApplyS
     )
 }
 
+fn validate_states(raft_state: &RaftLocalState, apply_state: &RaftApplyState) -> Result<()> {
+    let last_index = raft_state.get_last_index();
+    let commit_index = raft_state.get_hard_state().get_commit();
+    let apply_index = apply_state.get_applied_index();
+    if last_index < commit_index {
+        return Err(box_err!("raft_state {:?} corrupted", raft_state));
+    }
+
+    if commit_index >= apply_index {
+        if commit_index >= apply_state.get_last_commit_index() {
+            return Ok(());
+        }
+        return Err(box_err!(
+            "raft log lost detected: {:?} {:?}",
+            raft_state,
+            apply_state
+        ));
+    }
+    if commit_index < apply_state.get_last_commit_index() || apply_index > last_index {
+        return Err(box_err!(
+            "raft state {:?} not match apply state {:?} and can't be recovered,",
+            raft_state,
+            apply_state
+        ));
+    }
```

Comment:
I don't understand. It should be valid when commit_index < last_commit_index.

Reply:
When `commit_index` is less then `last_commit_index`, there must be data lost. It's expected to panic instead of letting error pass silently.